### PR TITLE
Fix dtype option of sum and prod

### LINKED ISSUE #1240: sum without upcast
`cupy.sum` and `cupy.prod` upcasts ints (or bool) to `int64` or `uint64`, to align with numpy.  This feature would be disabled with `x.sum(dtype=x.dtype)` but not supported in cupy.
```
>>> x = cupy.arange(3).astype(cupy.int16)
>>> x.sum(dtype=x.dtype)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "cupy/core/core.pyx", line 1139, in cupy.core.core.ndarray.sum
  File "cupy/core/core.pyx", line 1147, in cupy.core.core.ndarray.sum
  File "cupy/core/reduction.pxi", line 222, in cupy.core.core.simple_reduction_function.__call__
  File "cupy/core/elementwise.pxi", line 698, in cupy.core.core._guess_routine
TypeError: Wrong type (<class 'numpy.int16'>) of arguments for cupy_sum
>>> cupy.__version__
'4.0.0'
```

### PR DESCRIPTION
**Pull request by toslunar**
This PR fixes the logic of `dtype` argument of `sum` and `prod`, so that it enables e.g.
`cupy.sum(cupy.arange(3).astype(cupy.int16), dtype=cupy.int32)`

close #1240.

### CODE REVIEW

**Comment by toslunar** [comment]
```
>>> np.core.umath.add.types
['??->?', 'bb->b', 'BB->B', 'hh->h', 'HH->H', 'ii->i', 'II->I', 'll->l', 'LL->L', 'qq->q', 'QQ->Q', 'ee->e', 'ff->f', 'dd->d', 'gg->g', 'FF->F', 'DD->D', 'GG->G', 'Mm->M', 'mm->m', 'mM->M', 'OO->O']
```

numpy just overwrites `dtype=None`.
https://github.com/numpy/numpy/blob/v1.14.3/numpy/core/src/umath/ufunc_object.c#L3848-L3870

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -3966,21 +3972,39 @@ _any = create_reduction_func(
 # Mathematical functions
 # -----------------------------------------------------------------------------
 
-_sum = create_reduction_func(
+_sum_auto_dtype = create_reduction_func(
     'cupy_sum',
-    ('?->l', 'B->L', 'h->l', 'H->L', 'i->l', 'I->L', 'l->l', 'L->L',
+    ('?->l', 'b->l', 'B->L', 'h->l', 'H->L', 'i->l', 'I->L', 'l->l', 'L->L',
```
Comment (kmaehashi) [inline cupy/core/core.pyx]: Could you explain the effect of this change?
(I think it makes sense to add this, but it seems there's no difference of behavior.)

File: tests/cupy_tests/math_tests/test_sumprod.py [hunk 2]
```diff
@@ -118,6 +118,16 @@ def test_sum_axes4(self, xp, dtype):
         a = testing.shaped_arange((20, 30, 40, 50), xp, dtype)
         return a.sum(axis=(0, 2, 3))
 
+    @testing.for_all_dtypes_combination(names=['src_dtype', 'dst_dtype'])
+    @testing.numpy_cupy_allclose()
+    def test_sum_dtype(self, xp, src_dtype, dst_dtype):
+        if not xp.can_cast(src_dtype, dst_dtype) \
+                or [xp.dtype(src_dtype).kind,
+                    xp.dtype(dst_dtype).kind].count('c') == 1:
```
Comment (kmaehashi) [inline tests/cupy_tests/math_tests/test_sumprod.py]: I guess it should be `src_dtype == xp.float16 and xp.dtype(dst_dtype).kind == 'c'`?

**Comment by kmaehashi** [comment]
Jenkins, test this please.

**Comment by kmaehashi** [comment]
LGTM.
Jenkins, test this please.

**Comment by kmaehashi** [comment]
LGTM!

**Pull request merged**